File: src/test/java/blynx/thymeleaf/compositiondialect/testcomponents/MagicHeadings.java
package blynx.thymeleaf.compositiondialect.testcomponents;

import blynx.thymeleaf.compositiondialect.CompositionComponent;
import blynx.thymeleaf.compositiondialect.CompositionComponentContext;

public record MagicHeadings(Integer level, CompositionComponentContext context) implements CompositionComponent {

    public MagicHeadings {
        // Read the parent's level BEFORE registering ourselves, so nested instances increment.
        if (level == null) {
            level = context.variable("parentMagicHeadings") instanceof MagicHeadings parent
                    ? parent.level() + 1
                    : 1;
        }
        context.setVariable("parentMagicHeadings", this);
    }
}

File: dialect-lib/src/test/kotlin/blynx/thymeleaf/compositiondialect/testcomponents/MagicHeadings.kt
package blynx.thymeleaf.compositiondialect.testcomponents

import blynx.thymeleaf.compositiondialect.CompositionComponent
import blynx.thymeleaf.compositiondialect.CompositionComponentContext

class MagicHeadings(context: CompositionComponentContext) : CompositionComponent(context) {
    val level: Int = ((context.variable("parentMagicHeadings") as? MagicHeadings)?.level ?: 0) + 1
    init { context.setVariable("parentMagicHeadings", this) }
}
